subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DomainObjectContextProjectStateHandler.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.transform;
+
+import org.gradle.api.Project;
+import org.gradle.api.internal.DomainObjectContext;
+import org.gradle.api.internal.artifacts.dsl.dependencies.ProjectFinder;
+import org.gradle.api.internal.project.ProjectState;
+import org.gradle.api.internal.project.ProjectStateRegistry;
+
+import javax.annotation.Nullable;
+
+public class DomainObjectContextProjectStateHandler {
+
+    private final ProjectStateRegistry projectStateRegistry;
+    private final DomainObjectContext domainObjectContext;
+    private final ProjectFinder projectFinder;
+
+    public DomainObjectContextProjectStateHandler(ProjectStateRegistry projectStateRegistry, DomainObjectContext domainObjectContext, ProjectFinder projectFinder) {
+        this.projectStateRegistry = projectStateRegistry;
+        this.domainObjectContext = domainObjectContext;
+        this.projectFinder = projectFinder;
+    }
+
+    @Nullable
+    public Project maybeGetOwningProject() {
+        if (domainObjectContext.getProjectPath() != null) {
+            return projectFinder.findProject(domainObjectContext.getProjectPath().getPath());
+        } else {
+            return null;
+        }
+    }
+
+    public boolean hasMutableProjectState() {
+        Project project = maybeGetOwningProject();
+        if (project != null) {
+            ProjectState projectState = projectStateRegistry.stateFor(project);
+            return projectState.hasMutableState();
+        }
+        return true;
```

Comment:
Maybe this is trivial, but why does no owning project mean having mutable project state?

Reply:
I think `hasMutableProjectState()` actually means that this thread is allowed to use the mutable project state, i.e. it has the lock for it. If there is no project, then there is nothing to modify. A better name for this method would be better.

Reply:
`ensureProjectStateIsMutable()`?